### PR TITLE
[GISEL] Fix bugs and clarify spec of G_EXTRACT_SUBVECTOR

### PR DESCRIPTION
**Pull request by michaelmaitland**
The implementation was missing the fact that `G_EXTRACT_SUBVECTOR` destination and source vector can be different types.

Also fix a bug in the MIR builder for `G_EXTRACT_SUBVECTOR` to generate the correct opcode.

Clarify the G_EXTRACT_SUBVECTOR specification.

### CODE REVIEW

**Comment by tschuett** [comment]
I agree with your change, but the machine verifier probably needs updates.

**Comment by michaelmaitland** [comment]
From [GISel Docs](https://llvm.org/docs/GlobalISel/GenericOpcode.html#g-extract-subvector)

> Extract a vector of destination type from the source vector. The index operand represents the starting index from which a subvector is extracted from the source vector.

> The index must be a constant multiple of the source vector’s minimum vector length. If the source vector is a scalable vector, then the index is first scaled by the runtime scaling factor. The indices extracted from the source vector must be valid indices of that vector. If this condition cannot be determined statically but is false at runtime, then the result vector is undefined.

From ISD::EXTRACT_SUBVECTOR:
>  Let the result type be T, then IDX represents the starting element number from which a subvector of type T is extracted. IDX must be a constant multiple of T's known minimum vector length. If T is a scalable vector, IDX is first scaled by the runtime scaling factor of T. Elements IDX through (IDX + num_elements(T) - 1) must be valid VECTOR indices. If this condition cannot be determined statically but is false at runtime, then the result vector is undefined. The IDX parameter must be a vector index constant type, which for most targets will be an integer pointer type.
> This operation supports extracting a fixed-width vector from a scalable vector, but not the other way around.

I don't see in here where we specify that scalable vector result needs to have same type as source.

It sounds like selection dag supports extracting fixed from scalable, which would mean that the source and destination are not the same type. But we don't yet specify whether we will permit this in GISel.

I would imagine that legalizers could be responsible for what different types they'd like to permit.

**Comment by tschuett** [comment]
"This operation supports extracting a fixed-width vector from a scalable vector, but not the other way around."
Then the types differ.

But extracting from  a scalable vector of pearls should return a scalable vector of pearls.

**Comment by michaelmaitland** [comment]
> But extracting from a scalable vector of pearls should return a scalable vector of pearls.

Maybe this should be enforced in the MV.

**Comment by tschuett** [comment]
We state "Extract a vector of destination type from the source vector."

**Comment by michaelmaitland** [comment]
@tschuett I have updated this patch to allow different types but enforce that scalable types must be the same. I also caught another related bug in the builder.

**Comment by tschuett** [comment]
Do we want to commit to the DAG comment on fixed-width vectors?

**Comment by michaelmaitland** [comment]
As I have written this patch, I have made no commitment to `Extract a vector of destination type from the source vector.` I'm content to revisit this once someone has a use case, unless you'd like to address it now.

**Comment by topperc** [comment]
For scalable vector input and scalable vector output, G_EXTRACT_SUBVECTOR should produce a vector with a smaller minimum number of elements than the input vector. The element types are the same, but the total vector size is smaller.

**Comment by michaelmaitland** [comment]
I have updated this patch to reflect this.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 1]
```diff
@@ -1778,10 +1778,9 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
       break;
     }
 
-    if (IndexOp.getImm() != 0 &&
-        SrcTy.getElementCount().getKnownMinValue() % IndexOp.getImm() != 0) {
-      report("Index must be a multiple of the source vector's minimum vector "
-             "length",
+    if (IndexOp.getImm() % DstTy.getElementCount().getKnownMinValue() != 0) {
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1781]: Please add the equivalent of this from the IR verifier

```
    // If this extraction is not the 'mixed' case where a fixed vector is        
    // extracted from a scalable vector, ensure that the extraction does not     
    // overrun the parent vector.                                                
    if (VecEC.isScalable() == ResultEC.isScalable()) {                           
      Check(IdxN < VecEC.getKnownMinValue() &&                                   
                IdxN + ResultEC.getKnownMinValue() <= VecEC.getKnownMinValue(),  
            "vector_extract would overrun.");                                    
    }  
```

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 2]
```diff
@@ -1778,10 +1778,9 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
...
+      report("Index must be a multiple of the destinations vector's minimum "
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1782]: destinations -> destination

File: llvm/lib/CodeGen/MachineVerifier.cpp (hunk 1 above)
Comment (tschuett) [reply llvm/lib/CodeGen/MachineVerifier.cpp:1781]: `mixed` is currently not documented/supported in GlobalIsel.

Comment (topperc) [reply llvm/lib/CodeGen/MachineVerifier.cpp]: Then the verifier needs to check that the scalable bits match.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 3]
```diff
@@ -1778,10 +1778,19 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
       break;
     }
 
-    if (IndexOp.getImm() != 0 &&
-        SrcTy.getElementCount().getKnownMinValue() % IndexOp.getImm() != 0) {
-      report("Index must be a multiple of the source vector's minimum vector "
-             "length",
+    int64_t Idx = IndexOp.getImm();
+    auto DstMinLen = DstTy.getElementCount().getKnownMinValue();
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1782]: I'm not sure `auto` here is consistent with llvm's guidelines on using `auto` does getKnownMinValue() return `unsigned` or `uint64_t`? I honestly don't know.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 4]
```diff
@@ -1778,10 +1778,19 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
       break;
     }
 
-    if (IndexOp.getImm() != 0 &&
-        SrcTy.getElementCount().getKnownMinValue() % IndexOp.getImm() != 0) {
-      report("Index must be a multiple of the source vector's minimum vector "
-             "length",
+    int64_t Idx = IndexOp.getImm();
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1781]: Should probably treat this as an unsigned value.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 5]
```diff
@@ -1778,10 +1778,20 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
...
+    uint64_t DstMinLen = DstTy.getElementCount().getKnownMinValue();
+    if (Idx % DstMinLen != 0) {
+      report("Index must be a multiple of the destination vector's minimum "
+             "vector length",
+             MI);
+      break;
+    }
+
+    uint64_t SrcMinLen = SrcTy.getElementCount().getKnownMinValue();
+    if (SrcTy.isScalable() && DstTy.isScalable() &&
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1791]: This applies for `!SrcTy.isScalable() && `DstTy.isScalable()` too.

The missing check is an independent check that `SrcTy.isScalable() == DstTy.isScalable()`.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 6]
```diff
@@ -1788,7 +1788,7 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
     }
 
     uint64_t SrcMinLen = SrcTy.getElementCount().getKnownMinValue();
-    if (SrcTy.isScalable() && DstTy.isScalable() &&
+    if (SrcTy.isScalable() == DstTy.isScalable() &&
```
Comment (topperc) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1791]: We've gone in a circle. We need a `report` if `SrcTy.isScalable() != DstTy.isScalable()`

Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineVerifier.cpp:1791]: Why? We haven’t specified the behavior for this case have we? I was thinking to have a follow up pr that synced the mixed case with the lang ref.

Comment (topperc) [reply llvm/lib/CodeGen/MachineVerifier.cpp:1791]: Well I was thinking you need to either disallow the mixed case or check that we don't have the unsupported mixed case of scalable dest and fixed source. So some check needs to exist.

Comment (michaelmaitland) [reply llvm/lib/CodeGen/MachineVerifier.cpp:1796]: Let's start out by not allowing it? We could probably change our mind down the line if we go in that direction (opposed to allowing it now and disallowing in the future).

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**